dowhy/causal_estimator.py

```diff
@@ -104,57 +86,68 @@ def __init__(
         :param kwargs: (optional) Additional estimator-specific parameters
         :returns: an instance of the estimator class.
         """
-        self._data = data
         self._target_estimand = identified_estimand
-        # Currently estimation methods only support univariate treatment and outcome
-        self._treatment_name = treatment
-        self._outcome_name = outcome[0]  # assuming one-dimensional outcome
-        self._control_value = control_value
-        self._treatment_value = treatment_value
+
         self._significance_test = test_significance
         self._effect_strength_eval = evaluate_effect_strength
-        self._target_units = target_units
-        self._effect_modifier_names = effect_modifiers
         self._confidence_intervals = confidence_intervals
-        self._bootstrap_estimates = None  # for confidence intervals and std error
-        self._bootstrap_null_estimates = None  # for significance test
-        self._effect_modifiers = None
+
         self.method_params = kwargs
         # Setting the default interpret method
         self.interpret_method = CausalEstimator.DEFAULT_INTERPRET_METHOD
 
         self.logger = logging.getLogger(__name__)
 
-        # Setting treatment and outcome values
-        if self._data is not None:
-            self._treatment = self._data[self._treatment_name]
-            self._outcome = self._data[self._outcome_name]
-
-        if self._effect_modifier_names:
-            # only add the observed nodes
-            self._effect_modifier_names = [
-                cname for cname in self._effect_modifier_names if cname in self._data.columns
-            ]
-            if len(self._effect_modifier_names) > 0:
-                self._effect_modifiers = self._data[self._effect_modifier_names]
-                self._effect_modifiers = pd.get_dummies(self._effect_modifiers, drop_first=True)
-                self.logger.debug("Effect modifiers: " + ",".join(self._effect_modifier_names))
-            else:
-                self._effect_modifier_names = None
-
         # Check if some parameters were set, otherwise set to default values
         self.num_null_simulations = num_null_simulations
         self.num_simulations = num_simulations
         self.sample_size_fraction = sample_size_fraction
         self.confidence_level = confidence_level
         self.num_quantiles_to_discretize_cont_cols = num_quantiles_to_discretize_cont_cols
         # Estimate conditional estimates by default
+        self.need_conditional_estimates = need_conditional_estimates
+
+        self._bootstrap_estimates = None
+        self._bootstrap_null_estimates = None
+
+    def set_data(self, data: pd.DataFrame, treatment_name: List[str], outcome_name: List[str]):
```

Comment:
Can we avoid this method? At least, I think we should not make it part of the contract of a `CausalEstimator`. If some implementation of `CausalEstimator` needs the data after `fit`, that specific implementation is still free to store it.

I think having such a method is too inviting to get into the business of very stateful objects. While an estimator is by definition stateful (it stores the artifacts coming out of training), I don't think there should be explicit methods for this kind of state management.

Reply:
Good point. I think we can avoid storing these data-based objects completely. 

For context, the earlier version had this code inside the init method of the estimator. But now the init method does not get access to the data, only fit method gets access to the data (as in sklearn API). That's why this method needed to be created. But it may not be needed.

Going forward, the three basic methods in an estimator's contract are `fit`, `effect` and `do` . It is not expected that fit and estimate effect should be called on the same data. So we probably don't need to store dataset inside self._data. Again we don't need to store the values of the treatment and outcome data. And the names of the treatment and outcome columns can be extracted from self.target_estimand. 

The only big change I see is that data_df will not be optional in estimate_effect method. Which is fine, because even sklearn API requires a dataset to be provided for the `predict` step.

@andresmor-ms what do you think? would removing this method lead to any problems?

Reply:
I'll take a look at this, I tried to do this in a previous version but found out that self._data was used in several places and decided to go with setting self._data and focus on getting the fit() to work correctly, if this becomes a big change, would you agree that it could go into another PR? @amit-sharma @petergtz

Reply:
Agree with @amit-sharma that ideally we wouldn't have to store this data at all. But to avoid delaying this PR too much, I'd be fine by simply renaming this method to `_set_data` to make it very explicit, that this is not part of the API of that class.